File: 3743.maximize-cyclic-partition-score.py
#
# @lc app=leetcode id=3743 lang=python3
#
# [3743] Maximize Cyclic Partition Score
#

# @lc code=start
from typing import List

class Solution:
    def maximumScore(self, nums: List[int], k: int) -> int:
        n = len(nums)
        if n == 1:
            return 0
        
        # Double the array to handle cyclic nature
        arr = nums + nums
        
        # Precompute ranges for all relevant subarrays
        range_arr = [[0] * (2 * n) for _ in range(2 * n)]
        for i in range(2 * n):
            mn = mx = arr[i]
            for j in range(i, min(i + n, 2 * n)):
                mn = min(mn, arr[j])
                mx = max(mx, arr[j])
                range_arr[i][j] = mx - mn
        
        best = 0
        
        # Try each starting position
        for start in range(n):
            # prev_dp[j] = max score for first j elements with m-1 partitions
            prev_dp = [-1] * (n + 1)
            prev_dp[0] = 0
            
            for m in range(1, k + 1):
                curr_dp = [-1] * (n + 1)
                for j in range(1, n + 1):
                    for i in range(j):
                        if prev_dp[i] >= 0:
                            val = prev_dp[i] + range_arr[start + i][start + j - 1]
                            curr_dp[j] = max(curr_dp[j], val)
                
                if curr_dp[n] >= 0:
                    best = max(best, curr_dp[n])
                
                prev_dp = curr_dp
                
                # Early termination if no valid states
                if all(x < 0 for x in prev_dp):
                    break
        
        return best
# @lc code=end

File: 3743.maximize-cyclic-partition-score.cpp
#
# @lc app=leetcode id=3743 lang=cpp
#
# [3743] Maximize Cyclic Partition Score
#
# @lc code=start
class Solution {
public:
    long long maximumScore(vector<int>& nums, int k) {
        int n = nums.size();
        if (n == 1) return 0;
        k = min(k, n);
        
        vector<int> prefixMax(n), prefixMin(n), suffixMax(n), suffixMin(n);
        prefixMax[0] = prefixMin[0] = nums[0];
        for (int i = 1; i < n; i++) {
            prefixMax[i] = max(prefixMax[i-1], nums[i]);
            prefixMin[i] = min(prefixMin[i-1], nums[i]);
        }
        suffixMax[n-1] = suffixMin[n-1] = nums[n-1];
        for (int i = n - 2; i >= 0; i--) {
            suffixMax[i] = max(suffixMax[i+1], nums[i]);
            suffixMin[i] = min(suffixMin[i+1], nums[i]);
        }
        
        vector<vector<long long>> rangeArr(n, vector<long long>(n));
        for (int l = 0; l < n; l++) {
            int curMax = nums[l], curMin = nums[l];
            for (int r = l; r < n; r++) {
                curMax = max(curMax, nums[r]);
                curMin = min(curMin, nums[r]);
                rangeArr[l][r] = curMax - curMin;
            }
        }
        
        const long long NEG_INF = LLONG_MIN / 2;
        
        vector<vector<vector<long long>>> dpSeg(n, vector<vector<long long>>(n, vector<long long>(k + 1, NEG_INF)));
        for (int l = 0; l < n; l++) dpSeg[l][l][1] = 0;
        
        for (int len = 2; len <= n; len++) {
            for (int l = 0; l + len - 1 < n; l++) {
                int r = l + len - 1;
                dpSeg[l][r][1] = rangeArr[l][r];
                for (int m = 2; m <= min(len, k); m++) {
                    for (int p = l + m - 2; p < r; p++) {
                        if (dpSeg[l][p][m-1] > NEG_INF) {
                            dpSeg[l][r][m] = max(dpSeg[l][r][m], dpSeg[l][p][m-1] + rangeArr[p+1][r]);
                        }
                    }
                }
            }
        }
        
        vector<vector<long long>> maxDpSeg(n, vector<long long>(n, NEG_INF));
        for (int l = 0; l < n; l++) {
            for (int r = l; r < n; r++) {
                int len = r - l + 1;
                for (int m = 1; m <= min(k - 1, len); m++) {
                    maxDpSeg[l][r] = max(maxDpSeg[l][r], dpSeg[l][r][m]);
                }
            }
        }
        
        long long ans = NEG_INF;
        for (int m = 1; m <= k; m++) {
            ans = max(ans, dpSeg[0][n-1][m]);
        }
        
        if (k >= 2) {
            for (int start = 2; start < n; start++) {
                for (int end = 0; end <= start - 2; end++) {
                    long long wrapMax = max(suffixMax[start], prefixMax[end]);
                    long long wrapMin = min(suffixMin[start], prefixMin[end]);
                    long long wrapRange = wrapMax - wrapMin;
                    
                    int midL = end + 1, midR = start - 1;
                    if (maxDpSeg[midL][midR] > NEG_INF) {
                        ans = max(ans, wrapRange + maxDpSeg[midL][midR]);
                    }
                }
            }
        }
        
        return ans;
    }
};
# @lc code=end